### ISSUE TITLE
All image previews are subtly desaturated

### ISSUE THREAD

**Issue by afontenot** [matrix-org/synapse#6021]
I've done my best to search for this issue, but couldn't find it previously reported. I hope it's okay to report this here (since it seems to be the correct product), even though I just use matrix.org + Riot.im instead of my own homeserver.

### Description

Basically, any image I upload is subtly modified in image previews, but not the full sized / lightbox version. This happens even if the preview and original are the same size. For a while I thought something was going wrong with color profile support in Firefox, but after looking more closely this doesn't seem to be the case (this occurs on images without color profiles, and the original image display fine in the lightbox).

### Steps to reproduce

- Upload any image
- Compare the preview image to the full image (I like to open both in tabs with the dev tools inspector for easy comparison)
- Note that the preview image is lighter / less saturated. I have verified this with ImageMagick's `identify` tool as well.

Image previews ought to be an accurate representation of the full sized image.

### Version information

- **Homeserver**: matrix.org (via Riot.im web app)

Tested today on Firefox 69.0 on an updated Arch Linux installation. Color management is enabled via [colord](https://www.freedesktop.org/software/colord/intro.html).

### Example images:
Bad (image preview):
![bad](https://user-images.githubusercontent.com/3870410/64734438-7fc2c100-d49b-11e9-8835-b8f4417e3862.png)

Good (original image):
![good](https://user-images.githubusercontent.com/3870410/64734503-9832db80-d49b-11e9-9adf-cbb4ac51718a.png)

**Comment by afontenot** [comment]
I don't think the difference is particularly clear in my images (above). Here are some samples that I created that should illustrate just how severely the previews are degraded.

Original image (created in GIMP, color bars are `#ff0000`, `#00ff00`, `#0000ff`) vs. Image preview, downloaded from Riot.im:

![rgbtest](https://user-images.githubusercontent.com/3870410/65559629-4e60e100-def0-11e9-9235-2f02c501a75a.png) ![rgbtest_preview](https://user-images.githubusercontent.com/3870410/65559644-5e78c080-def0-11e9-9eb8-4d72eee4ac74.png)

The full image, downloaded from Riot.im, was byte-for-byte the same as the original that I uploaded.

Can anyone else confirm this problem?

**Comment by richvdh** [comment]
I'm afraid I can't immediately reproduce this. I uploaded your original colour-bars image [link](https://matrix.sw1v.org/_matrix/media/r0/download/sw1v.org/MHqZpocSZaboFybuWCpWKNkA) and the [thumbnail](https://matrix.sw1v.org/_matrix/media/r0/thumbnail/sw1v.org/MHqZpocSZaboFybuWCpWKNkA?width=800&height=600) looks identical to me.

However, I would imagine it depends on how you have configured thumbnails on your server. What do you have `dynamic_thumbnails` and `thumbnail_sizes` set to?

It might also be a problem with your version of `pillow`. Have you tried updating it?

**Comment by afontenot** [comment]
> I'm afraid I can't immediately reproduce this. I uploaded your original colour-bars image [link](https://matrix.sw1v.org/_matrix/media/r0/download/sw1v.org/MHqZpocSZaboFybuWCpWKNkA) and the [thumbnail](https://matrix.sw1v.org/_matrix/media/r0/thumbnail/sw1v.org/MHqZpocSZaboFybuWCpWKNkA?width=800&height=600) looks identical to me.
> 
> However, I would imagine it depends on how you have configured thumbnails on your server. What do you have `dynamic_thumbnails` and `thumbnail_sizes` set to?
> 
> It might also be a problem with your version of `pillow`. Have you tried updating it?

This is on Riot.im.

**Comment by richvdh** [comment]
oh sorry, I missed that you had said you were using the matrix.org homeserver.

**Comment by ara4n** [comment]
this really sounds like a colorimetric profile problem - either that the profile is being stripped out in the thumbnail, or the thumbnailing is assuming a different source colourspace to the target colourspace when it thumbnails (e.g. 'raw' RGB to sRGB).

Can you share the URL of your test image & the test thumbnail of it from the matrix.org HS's media repo please, so I can see how my browser renders them, and what imagemagick makes of them?  I don't trust the images uploaded into Github and would rather see the originals.

**Comment by richvdh** [comment]
I'm also unable to reproduce it on matrix.org :/. Again, the [thumbnail](https://matrix.org/_matrix/media/r0/thumbnail/matrix.org/HawArkvTgDwXolUKwkcBwcGz?width=800&height=600) looks identical to the original to me :/

**Comment by afontenot** [comment]
> this really sounds like a colorimetric profile problem - either that the profile is being stripped out in the thumbnail, or the thumbnailing is assuming a different source colourspace to the target colourspace when it thumbnails (e.g. 'raw' RGB to sRGB).
> 
> Can you share the URL of your test image & the test thumbnail of it from the matrix.org HS's media repo please, so I can see how my browser renders them, and what imagemagick makes of them? I don't trust the images uploaded into Github and would rather see the originals.

Okay, so I've made an interesting discovery. This only happens on encrypted chats (all my ongoing chats are encrypted). I didn't discover this until I had to try to extract a URL to send you, but in an encrypted chat you don't get a URL for images, you get a `blob:` with some UUID. So unfortunately I can't send you a link.

This is probably happening because my device is color managed; presumably you have the browser itself do the thumbnailing in Riot, since (I assume) the server doesn't have access to the image? When I upload the image, I see the modified thumbnail in all other clients, including ones that aren't color managed. So I suppose something like this is happening:

1. Riot (web) uploads the real image, which can be downloaded and verified byte for byte.
2. To create the thumbnail, Riot (web) sticks the raw image data into a `canvas` or something, and (correctly) color manages the output of the canvas according to my display profile. It then snapshots the canvas into a thumbnail, but that captures the display profile corrected output, not the original colors. It uploads that modified version as the thumbnail image.

So basically, Riot (web) needs to downscale the image intelligently itself, instead of relying on the browser's rendering (which is color corrected on any managed display).

This is just a hypothesis; I'm not familiar with Riot's code.

**Comment by richvdh** [comment]
ohhh, in which case, this is a client problem. 

The fact that it only happens in encrypted rooms is critical, because it means the thumbnailing happens on the client rather than the server.

Please can you raise the bug in https://github.com/vector-im/riot-web instead? (bugs in matrix-react-sdk tend to get filed under riot-web, since that's where they end up manifesting)

**Comment by ara4n** [comment]
the hypothesis sounds correct ftr. i'm a bit surprised that we'd be capturing display profile corrected output from a canvas which is not connected to the DOM though... let's pick it up over on the riot-web repo.

**Comment by afontenot** [comment]
> the hypothesis sounds correct ftr. i'm a bit surprised that we'd be capturing display profile corrected output from a canvas which is not connected to the DOM though... let's pick it up over on the riot-web repo.

OK, I'm filing a bug there now. Will link when I'm done.
